scripts/metadata_comparison/metadata_comparison/extractor.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python3
+#
+# extractor.py
+#
+# Purpose: Read workflow metadata from Cromwell, and all metadata for its jobs,
+# and upload it to a GCS bucket
+#
+# Usage: python3 extractor.py <GCS path> <workflowId> [<workflowId2> [...]]
+#
+# Python Prereqs (at least, the ones which I needed to manually install... YMMV):
+#
+#   * pip3 install --upgrade requests
+#   * pip3 install --upgrade google-api-python-client
+#   * pip3 install --upgrade google-cloud
+#   * pip3 install --upgrade google-cloud-storage
+#
+# Remember to login to create application default credentials before use:
+#   % gcloud auth application-default login
+
+import argparse
+import re
+import json
+import requests
+from google.cloud import storage
+import google.auth
+from googleapiclient.discovery import build as google_client_build
+import logging as log
+
+logger = log.getLogger('metadata_comparison.extractor')
+
+def set_log_verbosity(verbose):
+    if verbose:
+        log.basicConfig(format='[%(asctime)s] [%(name)s] %(message)s', level=log.INFO)
+    else:
+        log.basicConfig(format='[%(asctime)s] [%(name)s] %(message)s', level=log.WARNING)
+        
+
+def quieten_chatty_imports():
+    log.getLogger('googleapiclient.discovery_cache').setLevel(log.ERROR)
+    log.getLogger('googleapiclient.discovery').setLevel(log.WARNING)
+
+
+def workflow_regex_validator(value):
+    """Makes sure that a value is a valid Cromwell workflow ID then returns the workflow ID"""
+    workflow_regex=re.compile('^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$')
+    if not workflow_regex.match(value):
+        msg = f'Invalid workflow ID {value}. Expected {workflow_regex.pattern}'
+        raise argparse.ArgumentTypeError(msg)
+    else:
+        return value
+
+
+def url_regex_validator(value):
+    """
+    Validates then extract the root of the Cromwell URL from the various URL strings which might be provided.
+    Deliberately flexible because it's tedious to remember which script requires which type of format.
+    eg:
+        'http://localhost' => 'http://localhost'
+        'http://localhost:8000' => 'http://localhost:8000'
+        'http://localhost:8000/' => 'http://localhost:8000'
+        'http://localhost:8000/api/workflows/' => 'http://localhost:8000'
+        'http://localhost:8000/custom/prefix/api/workflows/' => 'http://localhost:8000/custom/prefix'
+    """
```

Comment:
from the comment it sounds like this should live in a more universally importable place

Reply:
I was hoping to avoid this in part 1, but since I've gone as far as making the directory structure... sure, this shouldn't be too hard 🤞